File: src/main/scala/jk_5/nailed/server/map/stat/NailedStatManager.scala
package jk_5.nailed.server.map.stat

import jk_5.nailed.api.map.stat._
import jk_5.nailed.server.map.NailedMap

import scala.collection.convert.wrapAsScala._

/**
 * No description given
 *
 * @author jk-5
 */
class NailedStatManager(private val map: NailedMap) extends StatManager {

  private val stats = map.mappack match {
    case null => new Array[Stat](0)
    case m => m.getMetadata.stats.map{s =>
      if(s.track == null || s.track.isEmpty) new ModifiableStat(s.name) else new SubscribedStat(s.name, s.track, s.attributes)
    }.toArray
  }

  def fireEvent(event: StatEvent){
    stats.foreach{
      case s: SubscribedStat if s.track == event.name =>
        var stop = false
        for(e <- event.attributes.entrySet() if !stop){
          if(s.attributes.get(e.getKey) != e.getValue){
            stop = true
          }
        }
        s.onEvent(event)
      case _ =>
    }
  }

  def getStat(name: String): Stat = this.stats.find(_.name == name).orNull
}

File: src/main/java/jk_5/nailed/server/map/stat/NailedStatManager.java
package jk_5.nailed.server.map.stat;

import com.google.common.collect.ImmutableList;
import jk_5.nailed.api.map.Map;
import jk_5.nailed.api.map.stat.*;
import jk_5.nailed.api.mappack.metadata.StatConfig;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Collections;
import java.util.List;

public class NailedStatManager implements StatManager {

    private final Map map;
    private final List<Stat> stats;

    public NailedStatManager(Map map) {
        this.map = map;
        if(map.mappack() != null){
            ImmutableList.Builder<Stat> builder = ImmutableList.builder();
            for (StatConfig config : map.mappack().getMetadata().stats()) {
                if(config.track() == null || config.track().isEmpty()){
                    builder.add(new ModifiableStat(config.name()));
                }else{
                    builder.add(new SubscribedStat(config.name(), config.track(), config.attributes()));
                }
            }
            stats = builder.build();
        }else{
            this.stats = Collections.emptyList();
        }
    }

    public void fireEvent(StatEvent event){
        for (Stat stat : stats) {
            if(stat instanceof SubscribedStat){
                SubscribedStat s = (SubscribedStat) stat;
                if(s.getTrack().equals(event.getName())){
                    boolean matches = true;
                    for (java.util.Map.Entry<String, String> entry : event.getAttributes().entrySet()) {
                        if(!s.getAttributes().get(entry.getKey()).equals(entry.getValue())){
                            matches = false;
                            break;
                        }
                    }
                    if(!matches){ //TODO: this check was missing. Is it really needed?
                        s.onEvent(event);
                    }
                }
            }
        }
    }

    @Nullable
    public Stat getStat(@Nonnull String name){
        for (Stat stat : this.stats) {
            if(stat.getName().equals(name)){
                return stat;
            }
        }
        return null;
    }
}
